docs/releasehistory.rst

```diff
@@ -7,6 +7,60 @@ Releases follow the ``major.minor.micro`` scheme recommended by `PEP440 <https:/
 * ``minor`` increments add features but do not break API compatibility
 * ``micro`` increments represent bugfix releases or improvements in documentation
 
+
+
+0.4.0 - SMIRNOFF Spec Update and Performance Optimization
+---------------------------------------------------------
+
+This update introduces the 0.3 SMIRNOFF spec.
+This spec upgrade is the result of discussions about how to handle the evolution of data and parameter types as further functional forms are added to the SMIRNOFF spec.
+
+The toolkit now contains functionality to "upgrade" the spec of a data source.
+These methods are called automatically when loading an 0.1 or 0.2-spec data source.
+This functionality allows the toolkit to continue to read 0.2-spec files, and also implements backwards-compatibility for 0.1-spec files.
+
+
+WARNING: The 0.1 SMIRNOFF spec did not contain fields for several energy-determining parameters that are exposed in the 0.2 spec.
+Thus, the 0.1-to-0.2 spec conversion process makes assumptions about the values that should be added for the newly-required fields.
+These assumptions are intended to reproduce the spec update process for the smirnoff99Frosst family of force fields.
+If you use the spec conversion functionality for other force field families, please carefully review the warning messages printed during spec conversion to ensure they are providing your desired behavior.
+
+
+This update also greatly improves performance when running ``create_openmm_system`` on large topologies.
+
+SMIRNOFF Spec Changes
```

Comment:
Also, link to the SMIRNOFF 0.3 specification.

Reply:
I worry we're going to run into trouble here where there will be confusion between the global SMIRNOFF specification version, the top-level XML tag version, and individual parameter level versions. I think we have to clarify to the reader how these three versions are going to be separately updated:
* Will each update or extension to the SMIRNOFF spec increment the spec number?
* Will this also increment the top-level XML tag version, or does that only change when we add a new aromaticity model or something?
* How are the individual parameter level version numbers to be incremented now?

If we can clarify all of this to the reader, it will help!

Reply:
## Here are my thoughts for what we could include in the release notes:

### We update the section versions when:
* A new section attribute is exposed or an existing one is removed
* A new section attribute value becomes accepted (for example, a new "named" potential type like  "harmonic" or "charmm")
* We change the behavior of a section, but nothing about its syntax
* We change a default value in a section, but nothing about its syntax

### We update the SMIRNOFF version when:
* We add a new supported aromaticity model
* We add a new top-level attribute (it's hard to think of an example of these... maybe we find a limitation with SMIRKS and create a `chemical_perception="SMIRKS1"` or `"SMIRKS2"` option that users can switch between)

### We update the XML version when:
* Not within the spec/Not applicable. This is out of our control. The modular nature of `ParameterIOHandler`s means that users are free to serialize in any format they like, as long as the proper data ends up getting passed to the ForceField constructor (currently the "SMIRNOFF data dict", which is sort of a loose end but will stabilize as time goes on). 

## Here are some unresolved questions we may need to solve
### What is the relationship between a parameter section and the SMIRNOFF tag?

I actually don't know, so I'm going to "think out loud" here.

The SMIRNOFF tag currently encodes the aromaticity model. Could we achieve the same behavior by having each parameter section have an aromaticity model? Yes, we could. 

So maybe, the idea of an enclosing tag in the SMIRNOFF spec indicates "a value that would otherwise need to be set repeatedly in the enclosed tags". In that sense, the top-level `aromaticity_model` attribute is implicitly a value that affects, for example, the `BondHandler`'s behavior. 

As a thought experiment, I could make a new ParameterHandler that applies parameter based on a different aromaticity model, and have it override the higher level aromaticity model. This is a weird example, but I couldn't come up with a better one -- The point is that it would seem to be arbitrary to tell a developer they _can't_ do that, just because we say so.

Note that the above "sections should be enclosed in other sections if the enclosing section affects the behavior of the enclosee" logic would have a consequence for @andrrizzi's question in https://github.com/openforcefield/openforcefield/issues/310#issuecomment-498221565

> First thing that comes to mind is that we might want to have the charge model tags being children tags of Electrostatics, which will essentially have 3 levels of indentation instead of 2.

It implies that the sections that describe atomic charges should be enclosed by the Electrostatics tag (since attributes like the cutoff influence the functional form). It also opens the question of "what if a person wanted both the Electrostatics and GBSA tags? What should enclose the charge-determining sections?". I don't know what we'd do in that situation.

And, in a similar vein, this will make things like VirtualSites more difficult, since they currently are in their own section, but will experience electrostatics and vdW forces. Should they be doubly-nested  in Electrostatics and vdW tags? Or should the VirtualSite definitions be duplicated in the enclosing Electrostatics and vdW sections?

So maybe this isn't the right view of section enclosure.

## Maybe the SMIRNOFF version encodes something about the underlying section structure

For example, in the 0.2 -> 0.3 transition, we switch from units-in-the-header to units-in-the-quantity. Well, I'd argue that even _that_ change could have been handled at the parameter section level. So one could argue that changing to the "SMIRNOFF 0.3 spec" doesn't actually mean anything -- it's the parameter sections that need to change their behavior! But, I guess the SMIRNOFF spec change here records a "rule change", because the SMIRNOFF 0.3 spec allows ParameterHandlers with their own versions.

## Is a parameter section's version truly just the `version=Y` attribute in its defining tag? Or is it really a tuple of `(SMIRNOFF_version=X, section_version=Y)`?

I think it has to be the tuple, because if a parameter section only works with SMIRKS1, but the SMIRNOFF tag sets SMIRKS2... then the parameter section _needs_ to know about this higher-level setting that didn't even exist when it was written. So I guess each parameter section will need to have a `MAX_SMIRNOFF_VERSION` tag. And probably a `MIN_SMIRNOFF_VERSION` too.

Reply:
@jchodera @andrrizzi the first part of the post above is relevant to merging this PR. The bottom part can continue to be discussed but isn't blocking.

Reply:
> - Will each update or extension to the SMIRNOFF spec increment the spec number?
> - Will this also increment the top-level XML tag version, or does that only change when we add a new aromaticity model or something?
> - How are the individual parameter level version numbers to be incremented now?

I think we can consider using the top-level XML tag version as the "global" version that is attached to the general rules for building a SMIRNOFF XML file (i.e. root tag syntax and that each subtag is interpreted as an interaction type). By separating the versions of the single tags we have effectively created 5-6 quite independent specs.

Reply:
This could work, but is it necessary? We would ideally like to have developers experiment with evolving individual parameter types in parallel. The "official" spec would have to contain a certain combination of these individual parameter type specs, and this can get complicated and would break true modularity.

Reply:
@j-wags: Could we migrate the general part of your post to a separate issue?

Reply:
Migrated general part of the post to #120

Reply:
> I think we can consider using the top-level XML tag version as the "global" version that defines to the general rules for building a SMIRNOFF XML document

Maybe I'm misunderstanding, but I think we're going the wrong direction by thinking about the XML version here. We will serialize in versions other than XML, so having the XML version have any bearing on the meaning of a SMIRNOFF file is probably bad. 

I'd argue that any "general rules" about structure need to be encoded by the SMIRNOFF version. I could be convinced that we need a second "structure" version, but I'm having trouble thinking broadly enough to know what sort of change that might encode.

Reply:
I think I expressed myself confusingly. What I was trying to say there is that I think we need only two types of version: the one specified in the SMIRNOFF root tag, which is associated to the general grammar of the document, and the children tag version. Possibly, there will be multiple children tag versions that are compatible with the same SMIRNOFF root tag version.

Reply:
I agree that that the code only needs to concern itself with the root tag and section tag versions!

For the future, I think we were just trying to figure out if the "official SMIRNOFF spec" should also have a global version that encodes the most up-to-date collection of these root and section tag versions, or if this is not necessary.

Reply:
I see, thanks for the clarification. So essentially this would be versioning a "recommended" combination of tags and the GitHub documentation page? I can see the convenience of grouping a collection of versions into a single number for citation purposes. On the other hand, adding yet another version could generate some confusion.

I tend towards not adding right now it simply because I think people will have to cite a full offxml file(s) anyway, which already implies a collection of versions so I don't see a strong use case to support the added confusion, but I don't have a strong opinion about it, and I'll be happy with any decision you and Jeff will settle on.